extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,390 @@
+            final int fragmentOffset = (int) (localPosition - (currFragmentIndex * s3Instructions.fragmentSize()));
+            currentFragment.position(fragmentOffset);
+            numBytesCopied = Math.min(currentFragment.remaining(), destination.remaining());
+            final int originalBufferLimit = currentFragment.limit();
+            currentFragment.limit(currentFragment.position() + numBytesCopied);
+            destination.put(currentFragment);
+            // Need to reset buffer limit, so we can read from the same buffer again in future
+            currentFragment.limit(originalBufferLimit);
+        }
+        position = localPosition + numBytesCopied;
+        return numBytesCopied;
+    }
+
+    /**
+     * If the provided {@link SeekableChannelContext} is {@link SeekableChannelContext#NULL}, this method creates and
+     * returns a new {@link S3ChannelContext} with a cache size of 1 to support a single read with no read ahead, and
+     * the caller is responsible to close the context after the read is complete. Else returns the provided
+     * {@link SeekableChannelContext} cast to {@link S3ChannelContext}.
+     */
+    private static S3ChannelContext getS3ChannelContextFrom(@NotNull final SeekableChannelContext channelContext) {
+        final S3ChannelContext s3ChannelContext;
+        if (channelContext == SeekableChannelContext.NULL) {
+            s3ChannelContext = new S3ChannelContext(1);
+        } else {
+            Assert.instanceOf(channelContext, "channelContext", S3ChannelContext.class);
+            s3ChannelContext = (S3ChannelContext) channelContext;
+        }
+        return s3ChannelContext;
+    }
+
+    private int fragmentIndexForByteNumber(final long byteNumber) {
+        return Math.toIntExact(byteNumber / s3Instructions.fragmentSize());
+    }
+
+    /**
+     * If not already cached in the context, sends an async request to fetch the fragment at the provided index and
+     * caches it in the context.
+     */
+    private void sendAsyncRequest(final int fragmentIndex,
+            @NotNull final S3ChannelContext s3ChannelContext) {
+        if (s3ChannelContext.getCachedFuture(fragmentIndex) != null) {
+            // We have a pending request for this fragment
+            return;
+        }
+        final int fragmentSize = s3Instructions.fragmentSize();
+        final long readFrom = (long) fragmentIndex * fragmentSize;
+        final long readTo = Math.min(readFrom + fragmentSize, size) - 1;
+        final String range = "bytes=" + readFrom + "-" + readTo;
+
+        final int numBytes = (int) (readTo - readFrom + 1);
+        final ByteBufferAsyncResponseTransformer asyncResponseTransformer =
+                new ByteBufferAsyncResponseTransformer(bufferPool.take(numBytes));
+        final CompletableFuture<ByteBuffer> future = s3AsyncClient.getObject(
+                builder -> builder.bucket(bucket).key(key).range(range),
+                asyncResponseTransformer);
```

Comment:
Type warning here w/o generic. Easy to fix, but also, I think it's better to use the explicit request object instead of the consumer based one wrt getObject:

```java
        final ByteBufferAsyncResponseTransformer<GetObjectResponse> asyncResponseTransformer =
                new ByteBufferAsyncResponseTransformer<>(bufferPool.take(numBytes));
        final CompletableFuture<ByteBuffer> future = s3AsyncClient.getObject(GetObjectRequest.builder()
                .bucket(bucket)
                .key(key)
                .range(range)
                .build(), asyncResponseTransformer);
```